src/libs/kata-types/src/dmverity.rs

```diff
@@ -0,0 +1,509 @@
+// Copyright (c) 2023 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::{bail, Context, Error, Result};
+use base64;
+use devicemapper::{DevId, DmFlags, DmName, DmOptions, DM};
+use serde_json;
+use std::convert::TryFrom;
+use std::path::Path;
+
+/// Configuration information for DmVerity device.
+#[derive(Debug, Clone, Eq, PartialEq, Default, Serialize, Deserialize)]
+pub struct DmVerityInfo {
+    /// Hash algorithm for dm-verity.
+    pub hashtype: String,
+    /// Root hash for device verification or activation.
+    pub hash: String,
+    /// Size of data device used in verification.
+    pub blocknum: u64,
+    /// Used block size for the data device.
+    pub blocksize: u64,
+    /// Used block size for the hash device.
+    pub hashsize: u64,
+    /// Offset of hash area/superblock on hash_device.
+    pub offset: u64,
+}
+
+/// Configuration information for DmVerity device.
+
+impl DmVerityInfo {
+    /// Validate configuration information for DmVerity device.
+    pub fn validate(&self) -> Result<()> {
+        match self.hashtype.to_lowercase().as_str() {
+            "sha256" => {
+                if self.hash.len() != 64 || hex::decode(&self.hash).is_err() {
+                    bail!(
+                        "Invalid hash value sha256:{} for DmVerity device with sha256",
+                        self.hash,
+                    );
+                }
+            }
+            "sha1" => {
+                if self.hash.len() != 40 || hex::decode(&self.hash).is_err() {
+                    bail!(
+                        "Invalid hash value sha1:{} for DmVerity device with sha1",
+                        self.hash,
+                    );
+                }
+            }
+            // TODO: support ["sha224", "sha384", "sha512", "ripemd160"];
+            _ => {
+                bail!(
+                    "Unsupported hash algorithm {} for DmVerity device {}",
+                    self.hashtype,
+                    self.hash,
+                );
+            }
+        }
+
+        if self.blocknum == 0 || self.blocknum > u32::MAX as u64 {
+            bail!("Zero block count for DmVerity device {}", self.hash);
+        }
+        if !Self::is_valid_block_size(self.blocksize) || !Self::is_valid_block_size(self.hashsize) {
+            bail!(
+                "Unsupported verity block size: data_block_size = {},hash_block_size = {}",
+                self.blocksize,
+                self.hashsize
+            );
+        }
+        if self.offset % self.hashsize != 0 || self.offset < self.blocksize * self.blocknum {
+            bail!(
+                "Invalid hashvalue offset {} for DmVerity device {}",
+                self.offset,
+                self.hash
+            );
+        }
+
+        Ok(())
+    }
+
+    fn is_valid_block_size(block_size: u64) -> bool {
+        for order in 9..20 {
```

Comment:
Please could you add a comment explaining why we check this range.

Reply:
Thanks, I have added the comment about the block size.